packages/core/src/service/services/product.service.ts

```diff
@@ -136,18 +136,24 @@ export class ProductService {
     }
 
     async findOneBySlug(ctx: RequestContext, slug: string): Promise<Translated<Product> | undefined> {
-        const translations = await this.connection.getRepository(ctx, ProductTranslation).find({
-            relations: ['base'],
-            where: { slug },
-        });
-        if (!translations?.length) {
-            return;
-        }
-        const bestMatch =
-            translations.find(t => t.languageCode === ctx.languageCode) ??
-            translations.find(t => t.languageCode === ctx.channel.defaultLanguageCode) ??
-            translations[0];
-        return this.findOne(ctx, bestMatch.base.id);
+        const qb = this.connection.getRepository(ctx, Product).createQueryBuilder('product');
+        FindOptionsUtils.applyFindManyOptionsOrConditionsToQueryBuilder(qb, { relations: this.relations });
+        // tslint:disable-next-line:no-non-null-assertion
+        FindOptionsUtils.joinEagerRelations(qb, qb.alias, qb.expressionMap.mainAlias!.metadata);
+        return qb
+            .innerJoin(ProductTranslation, 'product_translation', 'product_translation.baseId=product.id')
+            .leftJoin('product.channels', 'channel')
+            .andWhere('product_translation.slug = :slug', { slug })
+            .andWhere('product_translation.languageCode = :languageCode', { languageCode: ctx.languageCode })
```

Comment:
Decided to implement slug tests for different translation cases. What result we except in each variant of test?

`describe('product query with translations', () => {
            it('by slug without translation arg', async () => {
                const translation = newTranslatedProduct.translations.filter(t => {
                    return t.languageCode === LanguageCode.en;
                })[0];
                const { product } = await adminClient.query<GetProductSimple.Query, GetProductSimple.Variables>(
                    GET_PRODUCT_SIMPLE,
                    { slug: translation.slug },
                );

                if (!product) {
                    fail('Product not found');
                    return;
                }
                expect(product.slug).toBe(translation.slug);
            });

            it('by slug de without translation arg', async () => {
                const translation = newTranslatedProduct.translations.filter(t => {
                    return t.languageCode === LanguageCode.de;
                })[0];
                const { product } = await adminClient.query<GetProductSimple.Query, GetProductSimple.Variables>(
                    GET_PRODUCT_SIMPLE,
                    { slug: translation.slug },
                );

                if (!product) {
                    fail('Product not found');
                    return;
                }
                expect(product.slug).toBe(translation.slug);
            });

            it('by slug with translation en', async () => {
                const translation = newTranslatedProduct.translations.filter(t => {
                    return t.languageCode === LanguageCode.en;
                })[0];
                const { product } = await adminClient.query<GetProductSimple.Query, GetProductSimple.Variables>(
                    GET_PRODUCT_SIMPLE,
                    { slug: translation.slug },
                    { languageCode: LanguageCode.en },
                );

                if (!product) {
                    fail('Product not found');
                    return;
                }
                expect(product.slug).toBe(translation.slug);
            });

            it('by slug de with translation en', async () => {
                const translation = newTranslatedProduct.translations.filter(t => {
                    return t.languageCode === LanguageCode.de;
                })[0];
                const { product } = await adminClient.query<GetProductSimple.Query, GetProductSimple.Variables>(
                    GET_PRODUCT_SIMPLE,
                    { slug: 'de-baked-potato' },
                    { languageCode: LanguageCode.en },
                );

                if (!product) {
                    fail('Product not found');
                    return;
                }
                expect(product.slug).toBe(translation.slug);
            });

            it('by slug with translation de', async () => {
                const translation = newTranslatedProduct.translations.filter(t => {
                    return t.languageCode === LanguageCode.en;
                })[0];
                const { product } = await adminClient.query<GetProductSimple.Query, GetProductSimple.Variables>(
                    GET_PRODUCT_SIMPLE,
                    { slug: 'en Baked Potato' },
                    { languageCode: LanguageCode.de },
                );

                if (!product) {
                    fail('Product not found');
                    return;
                }
                expect(product.slug).toBe(translation.slug);
            });

            it('by slug de with translation de', async () => {
                const translation = newTranslatedProduct.translations.filter(t => {
                    return t.languageCode === LanguageCode.de;
                })[0];
                const { product } = await adminClient.query<GetProductSimple.Query, GetProductSimple.Variables>(
                    GET_PRODUCT_SIMPLE,
                    { slug: 'de-baked-potato' },
                    { languageCode: LanguageCode.de },
                );

                if (!product) {
                    fail('Product not found');
                    return;
                }
                expect(product.slug).toBe(translation.slug);
            });
        });`
        
Test results for stable old logic:
`Error: expect(received).toBe(expected) // Object.is equality

Expected: "de-baked-potato"
Received: "en-baked-potato"
<Click to see difference>


    at Object.<anonymous> (/Users/artem.danilov/WebstormProjects/vendure/packages/core/e2e/product.e2e-spec.ts:496:38)
    at processTicksAndRejections (internal/process/task_queues.js:93:5)

Error: expect(received).toBe(expected) // Object.is equality

Expected: "de-baked-potato"
Received: "en-baked-potato"
<Click to see difference>


    at Object.<anonymous> (/Users/artem.danilov/WebstormProjects/vendure/packages/core/e2e/product.e2e-spec.ts:530:38)
    at processTicksAndRejections (internal/process/task_queues.js:93:5)

Failed: "Product not found"
Error: Failed: "Product not found"
    at Env.fail (/Users/artem.danilov/WebstormProjects/vendure/node_modules/jest-jasmine2/build/jasmine/Env.js:722:61)
    at fail (/Users/artem.danilov/WebstormProjects/vendure/node_modules/jest-jasmine2/build/jasmine/jasmineLight.js:156:23)
    at Object.<anonymous> (/Users/artem.danilov/WebstormProjects/vendure/packages/core/e2e/product.e2e-spec.ts:544:21)
    at processTicksAndRejections (internal/process/task_queues.js:93:5)
`

![image](https://user-images.githubusercontent.com/8737418/113260286-bf980600-92d6-11eb-8e50-3dd117a79b8b.png)

Reply:
That's a good idea to make more thorough tests. So the basic idea is that the query should always return results in the current active language. In these tests this will be en by default, unless you specify otherwise with the query string `languageCode: LanguageCode.de`. This is why those 3 tests are failing - even though you are querying with a de slug, the results will still be in en.

Reply:
Ok.

When querying "by slug de without translation arg" or 'by slug de with translation en' we receive now "en-baked-potato"? Is that correct or we should get None? (Now it works because of fourth row in
```javascript
 const bestMatch =
            translations.find(t => t.languageCode === ctx.languageCode) ??
            translations.find(t => t.languageCode === ctx.channel.defaultLanguageCode) ??
            translations[0];
``` )

When querying "by slug en with translation de" we receive now None. Is that right or we should receive 'de-baked-potato'?

Reply:
> When querying "by slug de without translation arg" or 'by slug de with translation en' we receive now "en-baked-potato"?

yes this is correct behaviour. We should always return results in the active language - en in this case as this is the channel's default language.

> When querying "by slug en with translation de" we receive now None

Ideally I think we should be able to return the de version here, but looking at the logic I see why this does not currently work. If you can make it work that way, this would be good. If not, it is acceptable to just keep the existing behaviour (returning none).

Reply:
Ok, will implement this logic
`
en slug | en translation | default translation => en slug
en slug | de translation => de slug
de slug | en translation | default translation => en slug
de slug | de translation | translation exists => de slug
de slug | de translation | translation not exists => en slug
`